File: 1415-the-k-th-lexicographical-string-of-all-happy-strings-of-length-n/1415-the-k-th-lexicographical-string-of-all-happy-strings-of-length-n.cpp
class Solution {
public:
    vector<string> ans;
    void backtrack(string curr, int n){
        //base case 
        if(curr.size() == n){
            ans.push_back(curr);
            return;
        }


        for(char ch: {'a','b','c'}){
            if(curr.empty() || curr.back() != ch){
                backtrack(curr+ ch, n);
            }
        }
    }
    string getHappyString(int n, int k) {
        backtrack("",n);
        if( k > ans.size()){
            return "";
        }
        return ans[k-1];        
    }
};

File: 1415-the-k-th-lexicographical-string-of-all-happy-strings-of-length-n/1415-the-k-th-lexicographical-string-of-all-happy-strings-of-length-n.java
class Solution {
    List<String> res = new ArrayList<>();
    public String getHappyString(int n, int k) {
        backtrack("", n);

        if (k > res.size()) return "";
        return res.get(k - 1);
    
    }
    private void backtrack(String curr, int n) {

        if (curr.length() == n) {
            res.add(curr);
            return;
        }

        char[] chars = {'a', 'b', 'c'};

        for (char ch : chars) {

            if (curr.length() == 0 || curr.charAt(curr.length() - 1) != ch) {
                backtrack(curr + ch, n);
            }

        }
}
}